rhp/v4/server.go

```diff
@@ -0,0 +1,1044 @@
+	}
+	formationTxn.FileContracts[0].RenterSignature = renterSigResp.RenterContractSignature
+
+	// add the renter signatures to the transaction
+	for i := range formationTxn.SiacoinInputs[:len(req.RenterInputs)] {
+		formationTxn.SiacoinInputs[i].SatisfiedPolicy = renterSigResp.RenterSatisfiedPolicies[i]
+	}
+
+	// add our signature to the contract
+	formationTxn.FileContracts[0].HostSignature = s.hostKey.SignHash(formationSigHash)
+
+	// add the renter's parents to our transaction pool to ensure they are valid
+	// and update the proofs.
+	if len(req.RenterParents) > 0 {
+		if _, err := s.chain.AddV2PoolTransactions(req.Basis, req.RenterParents); err != nil {
+			return errorBadRequest("failed to add formation parents to transaction pool: %v", err)
+		}
+	}
+
+	// get the full updated transaction set
+	basis, formationSet, err := s.chain.V2TransactionSet(basis, formationTxn)
+	if err != nil {
+		return fmt.Errorf("failed to get transaction set: %w", err)
+	} else if _, err = s.chain.AddV2PoolTransactions(basis, formationSet); err != nil {
+		return errorBadRequest("failed to broadcast formation transaction: %v", err)
+	}
+	s.syncer.BroadcastV2TransactionSet(basis, formationSet)
+
+	// add the contract to the contractor
+	err = s.contractor.AddV2Contract(TransactionSet{
+		Transactions: formationSet,
+		Basis:        basis,
+	}, Usage{
+		RPCRevenue: settings.Prices.ContractPrice,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to add contract: %w", err)
+	}
+
+	// send the finalized transaction set to the renter
+	return rhp4.WriteResponse(stream, &rhp4.RPCFormContractThirdResponse{
+		Basis:          basis,
+		TransactionSet: formationSet,
+	})
+}
+
+func (s *Server) handleRPCRenewContract(stream net.Conn) error {
+	var req rhp4.RPCRenewContractRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	}
+
+	// validate prices
+	prices := req.Prices
+	if err := prices.Validate(s.hostKey.PublicKey()); err != nil {
+		return fmt.Errorf("price table invalid: %w", err)
+	}
+
+	// get the settings and update the prices to match the signed prices
+	settings := s.settings.RHP4Settings()
+	settings.Prices = prices
+	// get the current tip
+	tip := s.chain.Tip()
+
+	// validate the request
+	if err := req.Validate(settings, tip); err != nil {
+		return rhp4.NewRPCError(rhp4.ErrorCodeBadRequest, err.Error())
+	}
+
+	// lock the existing contract
+	state, unlock, err := s.lockContractForRevision(req.Renewal.ContractID)
+	if err != nil {
+		return fmt.Errorf("failed to lock contract %q: %w", req.Renewal.ContractID, err)
+	}
+	defer unlock()
+
+	// validate challenge signature
+	existing := state.Revision
+	if !req.ValidChallengeSignature(existing) {
+		return errorBadRequest("invalid challenge signature")
+	}
+
+	renewal := rhp4.NewRenewal(existing, prices, req.Renewal)
+
+	// create the renewal transaction
+	renewalTxn := types.V2Transaction{
+		MinerFee: req.MinerFee,
```

Comment:
I noticed that when we validate `req.MinerFee`, we only check that the fee is > 0. In RHP3 we have fields on the price table to make sure we agree on a minimum.

My thought is whether we can't just eliminate that discrepancy altogether now that we no longer sign the full txn and only the contract.

That means, the renter funds the `ContractPrice` but doesn't send a miner fee. Instead, the host just pays for the miner fee. If the renter thinks the amount the host paid isn't enough, it can then just add another input and increase the fees right on its end right? 

cc @lukechampine

Reply:
Isn't there still a scenario where the host rejects the formation because the renter didn't pay a high enough contract price to broadcast the transaction.

Reply:
Maybe we can expose the not enough fees error in the tpool, have the host check for that error and accept the contract, the rebroadcast code automatically handles it later when (hopefully) fees are lower.